Adds sort to command usage help output

trevorlinton opened the pull request:
When showing the usage (if no command is available) the order of commands isn't sorted, this might be a undesirable and could be put into a config setting, but I'd like to see the tolerance to default to sorted rather than as-defined.

bcoe commented:
@trevorlinton the nice thing about not applying sorting, is folks can control the ordering of commands by definition order.

my concern with this change is it would break a lot of downstream consumers of yargs, and would need to be a breaking change.

trevorlinton commented:
can I add it as an optional config (disabled by default?) the particular issue I'm under is I don't have control with the timing when new commands are registered, so the ordering can at times be random.

bcoe commented:
@trevorlinton that's an interesting requirement; I'd be open to this, I've been chatting with @cspotcode about adding the ability to configure compiler settings with a flag like `.yargsConfig` which would pass config to the `yargs-parser`; what if we allowed certain yargs settings in the non-parser layer to be configured with this too, e.g., `sort-commands` which defaults to `false`.

FYI, I think you can use `String.localeCompare()` to perform the sortinng.

bcoe commented:
@trevorlinton see https://github.com/yargs/yargs/pull/1262, I'm comfortable with `sort-commands` being a configuration option that can be passed to yargs via `parserConfiguration`.

trevorlinton commented:
@evocateur @bcoe I added in the `sort-commands` based on the work done in #1262 is there a common branch I can do a PR/rebase into on yargs? otherwise I'm going to be pulling in all the commits/changes from #1262 and our merges become very similar (or, alternatively, merge #1262 first and leave this be until then).

trevorlinton commented on lib/usage.js:
```diff
@@ -34,7 +34,6 @@ module.exports = function usage (yargs, y18n) {
   let failureOutput = false
   self.fail = function fail (msg, err) {
     const logger = yargs._getLoggerInstance()
-
```
fixed, sorry.

trevorlinton commented on yargs.js:
```diff
@@ -1147,7 +1147,6 @@ function Yargs (processArgs, cwd, parentRequire) {
       if (err instanceof YError) usage.fail(err.message, err)
       else throw err
     }
-
```
fixed

trevorlinton commented on lib/usage.js:
```diff
@@ -204,7 +203,7 @@ module.exports = function usage (yargs, y18n) {
       const parentCommands = context.commands.length ? `${context.commands.join(' ')} ` : ''
 
-      commands.forEach((command) => {
+      commands.sort((a, b) => a[0] > b[0] ? 1 : -1).forEach((command) => {
```
updated

evocateur replied:
Thanks!

bcoe commented:
@trevorlinton I'll get #1262 landed ASAP, and then you can rebase against `master` 👍

bcoe commented:
@trevorlinton thanks for your patience, if you don't mind rebasing I'll give this a final review 👍

one note, we'll want to make sure we document the new API method appropriately.

trevorlinton commented:
@bcoe I've rebased on master, you can review the file changes, it may be prudent to squash this merge considering all of the conflicts/rebasing that occurred by merging in a unmerged PR and diverging. I've added some notes in api.md on the config option, but I'm not sure how we should document this properly, I'm hoping what's there is good enough or now.

bcoe commented on docs/api.md:
```diff
@@ -1151,8 +1151,13 @@ for details of this object
 ------------
-`parserConfiguration()` allows you to configure yargs. See [yargs-parser's configuration](https://github.com/yargs/yargs-parser#configuration) for
-settings specific to `yargs-parser`.
+`parserConfiguration()` allows you to configure yargs-parser. 
```
let's make this:

"`parserConfiguration()` allows you to configure advanced yargs features:"

trevorlinton replied:
done.

bcoe commented on docs/api.md:
```diff
@@ -1151,8 +1151,13 @@ for details of this object
-settings specific to `yargs-parser`.
+`parserConfiguration()` allows you to configure yargs-parser. 
+
+An object of options should be passed in.
```
let's change this to:

`obj` accepts the following configuration options:

trevorlinton replied:
done.

trevorlinton commented:
@bcoe I've made the changes to the docs requested.

bcoe commented:
@trevorlinton 👍 thank you; I'm excited that we now have a way to configure edge-case output options for folks (just like we do with the parser).

bcoe commented:
@trevorlinton please try:

`npm i yargs@next`

and let me know if it does the trick for you.

bcoe commented:
this is now available in `yargs@13.1.0`.
